src/box/lua/schema.lua

```diff
@@ -3593,30 +3593,84 @@ local function role_check_grant_revoke_of_sys_priv(priv)
     end
 end
 
-box.schema.role.grant = function(user_name, ...)
+local function box_schema_role_grant_do(user_name, ...)
     local uid = role_resolve(user_name)
     if uid == nil then
         box.error(box.error.NO_SUCH_ROLE, user_name)
     end
     role_check_grant_revoke_of_sys_priv(...)
     return grant(uid, user_name, ...)
 end
-box.schema.role.revoke = function(user_name, ...)
+
+local function box_schema_role_revoke_do(user_name, ...)
     local uid = role_resolve(user_name)
     if uid == nil then
         box.error(box.error.NO_SUCH_ROLE, user_name)
     end
     role_check_grant_revoke_of_sys_priv(...)
     return revoke(uid, user_name, ...)
 end
-box.schema.role.info = function(role_name)
+
+local function box_schema_role_info_do(role_name)
     local rid = role_resolve(role_name)
     if rid == nil then
         box.error(box.error.NO_SUCH_ROLE, role_name)
     end
     return info(rid)
 end
 
+-- Wrap a function into transaction if none is active.
+local function atomic_wrapper(func)
+    return function(...)
+        -- No reason to start a transaction of one is active already.
+        if box.is_in_txn() then
+            return func(...)
+        else
+            return box.atomic(func, ...)
+        end
+    end
+end
+
+box.schema.space.create = atomic_wrapper(box_schema_space_create_do)
```

Comment:
Having this list of functions distanced from the implementation makes the code difficult to follow. Why not drop the temporary functions and use the atomic wrapper right at function definition?

```diff
diff --git a/src/box/lua/schema.lua b/src/box/lua/schema.lua
index 87f931ca4353..8fbec7d76765 100644
--- a/src/box/lua/schema.lua
+++ b/src/box/lua/schema.lua
@@ -717,7 +717,7 @@ local function check_space_type(space_type)
 end
 
 box.schema.space = {}
-box.schema.space.create = function(name, options)
+box.schema.space.create = atomic_wrapper(function(name, options)
     check_param(name, 'name', 'string')
     local options_template = {
         if_not_exists = 'boolean',
@@ -791,7 +791,7 @@ box.schema.space.create = function(name, options)
 
     feedback_save_event('create_space')
     return box.space[id], "created"
-end
+end)
 
 -- space format - the metadata about space fields
 function box.schema.space.format(id, format)
```

Reply:
This makes the code a bit less accurate to me, but that's a matter of taste.

Reply:
With a separate code block wrapping all DDL functions in `atomic_wrapper` one by one, one can't understand whether a particular function is atomic by just looking at its definition - they'd have to browse to a different place in the code, which isn't really convenient. The `atomic_wrapper` looks like a function decorator to me so better use it at a function definition IMO.

It'd be a different story if we wrapped *all* functions in this wrapper *automatically* but AFAIU we can't do that because not all functions need the wrapping.

Reply:
Fair enough, fixed.

It turns out there's only 9 functions to be wrapped, other are either read-only, singlewritestament or non-DDL.